accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -214,36 +248,253 @@ public void forCommittedInEpoch(Ranges ranges, long epoch, Consumer<Command> con
+                        rangeCommands.putIfAbsent(new RangeAndTxnId(range, command.txnId()), command);
+            }
+        }
+
+        @Override
+        public void register(Seekable keyOrRange, Ranges slice, Command command)
         {
-            switch (keysOrRanges.kindOfContents()) {
+            switch (keyOrRange.domain())
+            {
+                default: throw new AssertionError();
+                case Key:
+                    forEach(keyOrRange, slice, forKey -> forKey.register(command));
+                    break;
+                case Range:
+                    rangeCommands.putIfAbsent(new RangeAndTxnId((Range)keyOrRange, command.txnId()), command);
+            }
+        }
+
+        public <O> O mapReduceForKey(Routables<?, ?> keysOrRanges, Ranges slice, BiFunction<CommandsForKey, O, O> map, O accumulate, O terminalValue)
+        {
+            switch (keysOrRanges.domain()) {
                 default:
                     throw new AssertionError();
                 case Key:
                     AbstractKeys<Key, ?> keys = (AbstractKeys<Key, ?>) keysOrRanges;
-                    return keys.stream()
-                            .filter(slice::contains)
-                            .map(this::commandsForKey)
-                            .map(map)
-                            .reduce(initialValue, reduce);
+                    for (Key key : keys)
+                    {
+                        if (!slice.contains(key)) continue;
+                        CommandsForKey forKey = commandsForKey(key);
+                        accumulate = map.apply(forKey, accumulate);
+                        if (accumulate.equals(terminalValue))
+                            return accumulate;
+                    }
+                    break;
                 case Range:
                     Ranges ranges = (Ranges) keysOrRanges;
-                    return ranges.slice(slice).stream().flatMap(range ->
-                            commandsForKey.subMap(range.start(), range.startInclusive(), range.end(), range.endInclusive()).values().stream()
-                    ).map(map).reduce(initialValue, reduce);
+                    Ranges sliced = ranges.slice(slice, Minimal);
+                    for (Range range : sliced)
+                    {
+                        for (CommandsForKey forKey : commandsForKey.subMap(range.start(), range.startInclusive(), range.end(), range.endInclusive()).values())
+                        {
+                            accumulate = map.apply(forKey, accumulate);
+                            if (accumulate.equals(terminalValue))
+                                return accumulate;
+                        }
+                    }
             }
+            return accumulate;
         }
 
-        public void forEach(Routables<?, ?> keysOrRanges, Ranges slice, Consumer<CommandsForKey> forEach)
+        public void forEach(Seekables<?, ?> keysOrRanges, Ranges slice, Consumer<CommandsForKey> forEach)
```

Comment:
Should the `foreach` be private now?

Reply:
Still public

Reply:
Sorry, must have lost it in one of the rebases